Track dependencies in Mods page

Trial97 opened the pull request:
To do:
- [x] get dependencies from mod
- [x] get dependencies from provider data(save them in packwiz)
- [x] add columns to sort the mods based on the number of dependencies and the number of required by
- [x] auto disable/enable dependencies
- [ ] a way to visualize the dependencies for each mod

This is more like a personal build, but I want to have a place to track it.
Stuff that blocks this PR(aka known issues): 
- Because the mod parsing was updated to fetch more information, the mods page loads slower 
- This doesn't tell you if you are missing stuff: use the `Verify Dependencies` button to fetch the deps from providers (some loaders do put libraries as dependency information, and I can't determine if it is a mod that needs to be installed or included in the jar)
- There are edge cases: AE2 QoL Recipes doesn't report that AE2 is a dependency in the mod file, so if you download AE2 from CF and AE2 QoL Recipes from Modrinth, the relation may not be shown (as there is no link between the CF project ID and Modrinth); other example what falls here is Iris and Sodium
- Incompatible mods will not be tackled in this PR

Other than that, feel free to test it out.

fixes #1844
fixes #280
fixes #2628

Trial97 committed:
feat: decode dependencies from mod jar

Signed-off-by: Trial97 <alexandru.tripon97@gmail.com>

Trial97 committed:
feat: store provider dependencies

Signed-off-by: Trial97 <alexandru.tripon97@gmail.com>

Trial97 committed:
feat: add requireBy and requires columns

Signed-off-by: Trial97 <alexandru.tripon97@gmail.com>

Trial97 committed:
feat: make dependencies auto disable/enable

Signed-off-by: Trial97 <alexandru.tripon97@gmail.com>

aurora-1-2-3 commented:
Any update on whether this is going to be merged?

Also trial, is there a way to try your builds early? I really need the new features and am eager to get started.

Mick235711 commented:
> Also trial, is there a way to try your builds early? I really need the new features and am eager to get started.

Yes, navigate to checks (either by the "successful checks" dropdown above the comment section, or by selecting the green checkmark before each commit number), choose one of "Build" check, in its page select Summary from the left pane, and scroll all the way to the bottom you can see a list of generated builds from this PR.

aurora-1-2-3 commented:
thank you so much, ima test all of their builds

getchoo commented on launcher/minecraft/mod/ModFolderModel.cpp:
```diff
@@ -240,3 +262,145 @@ void ModFolderModel::onParseSucceeded(int ticket, QString mod_id)
+    auto findById = [](QList<Mod*> mods, QString id) -> Mod* {
+        auto found = std::find_if(mods.begin(), mods.end(), [id](Mod* m) { return m->mod_id() == id; });
+        return found != mods.end() ? *found : nullptr;
+    };
```
This should probably just be made into a function, as it's copy-pasted from above

getchoo commented on launcher/ui/pages/instance/ModFolderPage.cpp:
```diff
@@ -133,7 +134,22 @@ void ModFolderPage::removeItems(const QItemSelection& selection)
+    auto affected = m_model->getAffectedMods(indexes, EnableAction::DISABLE);
+    if (!affected.isEmpty()) {
+        auto response = CustomMessageBox::selectable(this, tr("Confirm Disable"),
+                                                     tr("The mods you are tring to remove are required by %1 mods.\n"
```
```suggestion
                                                     tr("The mods you are tring to disable are required by %1 mods.\n"
```

getchoo commented on launcher/ui/widgets/InfoFrame.cpp:
Really wanted to get both of these boxes to show at the same time, but I actually couldn't find a mod that had its own (correctly defined) dependencies that was depended on by others lol

Trial97 replied:
![image](https://github.com/user-attachments/assets/df960585-b956-4670-9a15-93d349a14459)
occultims

getchoo reviewed (commented):
LGTM otheriwse

I think it'd be cool to flesh the current requirements/required by boxes into a full properties-esc dialog, but that'd definitely need some features (like idk, file size, more author info, blah blah) and can be done later

Thanks!

Trial97 committed:
feat: display mod dependencies

Signed-off-by: Trial97 <alexandru.tripon97@gmail.com>

TheKodeToad commented on launcher/minecraft/mod/ModFolderModel.cpp:
```diff
@@ -48,25 +48,34 @@
+                                   "Minecraft Versions", "Release Type", "Requires", "Required by" });
+    m_column_names_translated =
+        QStringList({ tr("Enable"), tr("Image"), tr("Name"), tr("Version"), tr("Last Modified"), tr("Provider"), tr("Size"), tr("Side"),
+                      tr("Loaders"), tr("Minecraft Versions"), tr("Release Type"), tr("Requires "), tr("Required by") });
```
```suggestion
                      tr("Loaders"), tr("Minecraft Versions"), tr("Release Type"), tr("Requires"), tr("Required By") });
```

TheKodeToad replied:
Not resolved

TheKodeToad commented on launcher/minecraft/mod/ModFolderModel.cpp:
```diff
@@ -195,6 +213,10 @@ QVariant ModFolderModel::headerData(int section, [[maybe_unused]] Qt::Orientatio
                     return tr("The release type.");
                 case SizeColumn:
                     return tr("The size of the mod.");
+                case RequiredByColumn:
+                    return tr("Number of mods for what this is needed.");
+                case RequiresColumn:
+                    return tr("Number of mods that this requires.");
```
I would prefer this:
```suggestion
                case RequiredByColumn:
                    return tr("For each mod, the number of other mods which depend on it.");
                case RequiresColumn:
                    return tr("For each mod, the number of other mods it depends on.");
```

Might be inconsistent but I want it to be easier to understand than time travel

TheKodeToad replied:
Also unresolved.

TheKodeToad commented on launcher/minecraft/mod/ModFolderModel.h:
```diff
@@ -85,8 +90,20 @@ class ModFolderModel : public ResourceFolderModel {
+    void onParseFinished();
+
+   private:
+    QHash<QString, QList<Mod*>> m_requiredBy;
+    QHash<QString, QList<Mod*>> m_requires;
```
Can you not get away with doing this to eliminate duplicates?
```suggestion
    QHash<QString, QSet<Mod*>> m_requiredBy;
    QHash<QString, QSet<Mod*>> m_requires;
```

Trial97 committed:
change to QSet to remove duplicates

Signed-off-by: Trial97 <alexandru.tripon97@gmail.com>

TheKodeToad commented:
FYI it is possible to toggle many mods at once by pressing enter

Trial97 commented:
> FYI it is possible to toggle many mods at once by pressing enter

Can you please elaborate? What if it is possible to toggle many mods at once by pressing enter? Does the dependency stuff not work? Or what?

TheKodeToad commented:
> // not sure how to handle a bunch of rows that are toggled(not even sure it is posible)

Trial97 committed:
fix toggle action

Signed-off-by: Trial97 <alexandru.tripon97@gmail.com>

Trial97 commented:
This should have fixed it: if you try to do the toggle thing, this will prioritize the enabled mod(aka will ensure that when a mod is enabled, the deps are also enabled, regardless if the user wanted to disable the dep by toggle). When enabling/disabling individual mods, it will behave normally

TheKodeToad commented on launcher/minecraft/mod/tasks/LocalModParseTask.cpp:
```diff
@@ -62,6 +62,36 @@ ModDetails ReadMCModInfo(QByteArray contents)
+            details.dependencies.append(dep);
+        };
+
+        if (firstObj.contains("requiredMods")) {
+            for (auto dep : firstObj.value("dependencies").toArray().toVariantList()) {
+                addDep(dep.toString());
+            }
+        } else if (firstObj.contains("dependencies")) {
+            for (auto dep : firstObj.value("dependencies").toArray().toVariantList()) {
+                addDep(dep.toString());
+            }
+        }
```
nitpick: just use json api directly
but you did also accidentally write dependencies in the first case :P
```suggestion
        if (firstObj.contains("requiredMods")) {
            for (auto dep : firstObj.value("requiredMods").toArray()) {
                addDep(dep.toString());
            }
        } else if (firstObj.contains("dependencies")) {
            for (auto dep : firstObj.value("dependencies").toArray()) {
                addDep(dep.toString());
            }
        }
```

TheKodeToad commented on launcher/minecraft/mod/tasks/LocalModParseTask.cpp:
```diff
@@ -199,6 +229,42 @@ ModDetails ReadMCModTOML(QByteArray contents)
+                    }
+                }
+            }
+        }
```
Oops I guess I reworked a whole function again...
Better null safety and less nesting
```suggestion
        if (!dependencies) {
            return;
        }
        for (auto& dep : *dependencies) {
            auto dep_table = dep.as_table();
            if (!dep_table) {
               continue;
            }

            auto modId = (*dep_table)["modId"].as_string();
            if (!modId) {
                continue;
            }
            if (modId->get() == "forge" || modId->get() == "neoforge" || modId->get() == "minecraft") {
                continue;
            }

            if (auto type = (*dep_table)["type"].as_string()) { // NeoForge format
                if (type->get() != "required") {
                    continue;
                }
            } else if (auto mandatory = (*dep_table)["mandatory"]) {
                if (!mandatory->get()) {
                    continue;
                }
            } else {
                continue;
            }

            details.dependencies.append(QString::fromStdString(modId->get()));
        }
```

Trial97 replied:
As much as I love the go approach here, I think it would just be better to keep a relatively smaller number of continue: use it in moderation
I applied this suggestion partially, as you would see: I consider the 3 continue at the end being a bit too much

TheKodeToad replied:
No null safety though for modID - it will segfault

TheKodeToad replied:
I also prefer the [key].as_whatever() approach as it is null safe without needing to check .contains 
(it's also used elsewhere in the same function)

TheKodeToad replied:
Also given the length of 
`details.dependencies.append(QString::fromStdString(modId->get()));`
it seems bad to duplicate it for every case

Trial97 replied:
Ok now it should address all the comments but without the extra continue

Trial97 committed:
rework parse dependency function

Signed-off-by: Trial97 <alexandru.tripon97@gmail.com>

Trial97 committed:
Update launcher/minecraft/mod/ModFolderModel.cpp

Co-authored-by: TheKodeToad <TheKodeToad@proton.me>
Signed-off-by: Alexandru Ionut Tripon <alexandru.tripon97@gmail.com>
Signed-off-by: Trial97 <alexandru.tripon97@gmail.com>

TayouVR reviewed (approved):
didn't fully analyze code, but tested it out. Generally I think its pretty decent.

Some small feedback points. (not blocking)
- the dependency panels replacing the name and description bugs me a little, especially since I usually have the window fullscreen and there is more than enough space. I think the layouting could use some tweaking
- when enabling a mod its dependencies get enabled too. I sometimes enable a mod, try to update it or start the game, it doesn't work (outdated mod, incompatible with another, etc.). Now when I disable it its dependencies stay enabled. Not sure if theres a good way to handle that tbh, so I wouldn't count this as a blocker. But I'm sure other people will complain about it later on as well.

Trial97 commented:
- for the first one: yes, sure, there may be a better way to display them, but increasing the size of the panel at the bottom of the list seems like it takes a bit too much space. Yes, on a bigger screen, it may look decent, but we should consider that smaller resolutions exist(let's not make the experience worse for some users who may not have bigger resolutions). I'm open to suggestions and discussions on how best to display them. 
- on second: there is no way the game would start if the dependencies are disabled. So if you update the mod and have deps, you would want the deps to be enabled; otherwise, the game will not start. And the launcher should try to ensure that all the deps are enabled. A possible fix would be just to make the dependency toggle configurable in some way

Thanks for considering my PR.

Trial97 committed:
Update launcher/minecraft/mod/ModFolderModel.cpp

Co-authored-by: TheKodeToad <TheKodeToad@proton.me>
Signed-off-by: Alexandru Ionut Tripon <alexandru.tripon97@gmail.com>
Signed-off-by: Trial97 <alexandru.tripon97@gmail.com>

TheKodeToad reviewed (changes requested):
- Disabling is no longer an easily reversible action. It should have confirmation if another mod is being disabled. (edit: same with enabling)
- What's up with the ux here lol
https://github.com/user-attachments/assets/a80d38ef-5bb5-4dbf-aeb6-0f20cc77523b
*When the arrow points east, the description is shown. When it points west, the dependencies are shown. Click on the arrow to rotate it by 180°.*
I don't think we should try to cram so much info into this pane, or at least use tabs (but the horizontal layout doesn't work well for this)
Like getchoo says:
> I think it'd be cool to flesh the current requirements/required by boxes into a full properties-esc dialog

Probably best to leave this out until this is added

Trial97 commented:
- yes it is not reversible, but  without enabling/disabling the  mods, the instance would not start
- I did not use tabs for the pane, so I do not take 20px of screen with just the tab name(I tried to make it so it doesn't use more space than it already does). I still do not like the idea of separating it into a different window, as I want users to go through the mods and see the deps without needing to click each mod, then a button. I can change the button name to represent the action better. Can we discuss alternatives other than opening a new modal to see the deps/required by?

TheKodeToad commented:
> yes it is not reversible, but without enabling/disabling the mods, the instance would not start

Yeah but there should still be a confirmation for toggling a mod if there's a chain reaction. If the user says no disable no mods, if yes disable all mods

TheKodeToad commented:
> I did not use tabs for the pane, so I do not take 20px of screen with just the tab name(I tried to make it so it doesn't use more space than it already does). I still do not like the idea of separating it into a different window, as I want users to go through the mods and see the deps without needing to click each mod, then a button. I can change the button name to represent the action better. Can we discuss alternatives other than opening a new modal to see the deps/required by?

How about showing a list on hover?
Edit: this implementation also does appear to make the box a lot larger

TheKodeToad commented:
Also the counting breaks really easily lol

[Screencast_20251015_121214.webm](https://github.com/user-attachments/assets/6b85202c-668e-41ce-b688-fd3530631eb3)

Trial97 committed:
add confirmation dialog

Signed-off-by: Trial97 <alexandru.tripon97@gmail.com>

Trial97 commented:
> Yeah but there should still be a confirmation for toggling a mod if there's a chain reaction. If the user says no disable no mods, if yes disable all mods

Added the mentioned dialog, has yes(will automatically do the stuff), No( it will only disable the selected), Cancel(will do nothing). Language can be improved, and I need to make Cancel as default instead of No.






> Also the counting breaks really easily lol

Should have addressed that(hopefully) 





> How about showing a list on hover?

I need a bit more time on this.  And as I did not try it, I have my doubts about whether it is better or not.

> Edit: this implementation also does appear to make the box a lot larger

Yes, it has a few additional pixels because I added that button there.

Trial97 committed:
update toggle message box

Signed-off-by: Trial97 <alexandru.tripon97@gmail.com>

Trial97 committed:
improve toggle message

Signed-off-by: Trial97 <alexandru.tripon97@gmail.com>

TheKodeToad commented on launcher/minecraft/mod/ModFolderModel.cpp:
```diff
@@ -238,3 +263,253 @@ void ModFolderModel::onParseSucceeded(int ticket, QString mod_id)
+        QString yesButton;
+        if (requiredToEnable.size() > 0 && requiredToDisable.size() > 0) {
+            title = tr("Confirm toggle");
+            message = tr("Toggling this mod(s) will cause changes to other mods.\n") +
```
```suggestion
            message = tr("Toggling these mod(s) will cause changes to other mods.\n") +
```

TheKodeToad commented on launcher/minecraft/mod/ModFolderModel.cpp:
```diff
@@ -238,3 +263,253 @@ void ModFolderModel::onParseSucceeded(int ticket, QString mod_id)
+        mod->setRequiredByCount(m_requiredBy[id].count());
+        mod->setRequiresCount(m_requires[id].count());
+        int row = m_resources_index[mod->internal_id()];
+        emit dataChanged(index(row), index(row, columnCount(QModelIndex()) - 1));
```
You shouldn't need to cause an update for every column and row
in the whole treeview

TheKodeToad replied:
@Trial97 It still looks like it updates it for all rows

Trial97 replied:
ok now I'm sure that the update only happens when the number changes

TheKodeToad commented on launcher/ui/pages/instance/ModFolderPage.cpp:
```diff
@@ -135,7 +136,22 @@ void ModFolderPage::removeItems(const QItemSelection& selection)
+    auto affected = m_model->getAffectedMods(indexes, EnableAction::DISABLE);
+    if (!affected.isEmpty()) {
+        auto response = CustomMessageBox::selectable(this, tr("Confirm Disable"),
+                                                     tr("The mods you are tring to disable are required by %1 mods.\n"
```
```suggestion
        auto response = CustomMessageBox::selectable(this, tr("Confirm Disable"),
                                                     tr("The mods you are trying to delete are required by %1 mods.\n"
```

TheKodeToad commented on launcher/minecraft/mod/ModFolderModel.cpp:
```diff
@@ -238,3 +263,253 @@ void ModFolderModel::onParseSucceeded(int ticket, QString mod_id)
+        if (requiredToEnable.size() > 0 && requiredToDisable.size() > 0) {
+            title = tr("Confirm toggle");
+            message = tr("Toggling this mod(s) will cause changes to other mods.\n") +
+                      tr("%n mod(s) will be enabled\n", "", requiredToEnable.size()) +
```
%n might not be a good idea right now as we don't have translation files for en_US so it will probably always appear with the awkward (s)

Trial97 replied:
Yeah, but the alternative also does this. So the text would be the same even if I use %n or %1 and the QString::arg function.

Trial97 committed:
Apply suggestions from code review

Co-authored-by: TheKodeToad <TheKodeToad@proton.me>
Signed-off-by: Alexandru Ionut Tripon <alexandru.tripon97@gmail.com>

Trial97 committed:
only emit changes on the modified collumns

Signed-off-by: Trial97 <alexandru.tripon97@gmail.com>

Trial97 committed:
only emit changes when the count changes

Signed-off-by: Trial97 <alexandru.tripon97@gmail.com>

TheKodeToad commented on launcher/ui/pages/instance/ModFolderPage.cpp:
```diff
@@ -135,7 +136,22 @@ void ModFolderPage::removeItems(const QItemSelection& selection)
+                                                     tr("The mods you are trying to delete are required by %1 mods.\n"
+                                                        "Do you want to disable them?")
+                                                         .arg(affected.length()),
+                                                     QMessageBox::Warning, QMessageBox::Yes | QMessageBox::No, QMessageBox::No)
```
Just a nitpick: this should probably have cancel too

TheKodeToad commented on launcher/ui/widgets/InfoFrame.cpp:
```diff
@@ -227,8 +227,7 @@ void InfoFrame::updateWithResourcePack(ResourcePack& resource_pack)
 
-void InfoFrame::updateWithDataPack(DataPack& data_pack)
-{
+void InfoFrame::updateWithDataPack(DataPack& data_pack) {
```
Wrong brace style

Trial97 committed:
add cancel to confirm disable dialog

Signed-off-by: Trial97 <alexandru.tripon97@gmail.com>

DioEgizio commented:
with this pr prism spams "   169.934 Warning: "Failed to read str property 'version' in mod metadata." (Packwiz::stringEntry:74)" in logs while installing a cf pack. it's especially bad in big packs like atm10, where 
<img width="458" height="81" alt="image" src="https://github.com/user-attachments/assets/4082e0e4-fb35-4ceb-861e-2a8ff1f00416" />

Trial97 commented:
@DioEgizio that is not related to this PR.
Here is the PR that removes that exact warning https://github.com/PrismLauncher/PrismLauncher/pull/4388

DioEgizio commented:
> @DioEgizio that is not related to this PR.
> Here is the PR that removes that exact warning https://github.com/PrismLauncher/PrismLauncher/pull/4388

interesting, because I couldn't reproduce that on develop

Trial97 committed:
fix packwiz log spam

Signed-off-by: Trial97 <alexandru.tripon97@gmail.com>
